### PR TITLE
feat: IATR cloud data backend 

### PR DESCRIPTION
**Pull request by warrensplayer**
- Closes https://github.com/cypress-io/cypress/issues/24440

### User facing changelog

N/A merging to feature branch

### Additional details

**Background**
This PR moves the logic for determining the "relevant" run for the new Debug page to the server and exposing it through a GraphQL query.  The server handles fetching the git hashes to send to the Cypress cloud and processing the runs returned to determine the appropriate run(s) for the front end.

**What is a "relevant" run?**
A "relevant" run for the Debug page is determined by looking up runs in a Cypress Cloud project that have a Git commit sha that matches one of the last 100 commit shas in the users's current development project.  If the current environment does not use Git, then the Debug feature will not be available for that project.  See the `RelevantRunsDataSource` to see the logic for how the "current" and "next" relevant runs are determined.

**Main Components**
* packages/data-context/src/sources/RelevantRunsDatasource.ts
  * Server data source that determines which CloudRuns match the local Git shas and should be used for the Debug page
* packages/data-context/src/sources/RelevantRunSoecsDatasource.ts
  * Server data source that pulls spec counts for the current and next runs.  Used by a poller that backs a new GQL subscription 
* packages/app/src/composables/useRelevantRun.ts
  * GQL query for querying for the relevant run order to use the runNumber in subsequent queries
* packages/app/src/composables/useTestsForDebug.ts
  * GQL query for finding the failed tests to use for a filter in the runner when filtering in debug mode 

**GQL Updates/Additions**
* CurrentProject
  * added `testsForReviewBySpec`
  * TODO: Need to complete resolver for `testsForReviewBySpec`.  Currently just a placeholder 
* Mutation
  * added `movetoNextRelevantRun` for front end to indicate that the cached run just be removed and the newest run used as the current run
* RelevantRun
  * New field type added for representing the relevant current and next run
* Subscription
  * added `relevantRuns` to allow for the front end to see changes to the current and next relevant runs
  * added `relevantRunSpecChange` to allow for the banner and page that are showing updating spec counts for RUNNING builds to update

**Dependency Updates**
* Added `mocha-junit-reporter` and `moch-multi-reporters` to the `data-context` package directly
  * This was required to fix a bug when a test would fail an assertion.  These reporters were being referenced from the root `package.json`.  They would make a call out to the peer dependency for `mocha` which is currently an older version at the root of the project. Various versions of mocha are used throughout the mono-repo that should be looked at. 

**Other Updates**
Updated the various parts of the Debug feature that was using hardcoded `runNumber`s to use the new `useRelevantRuns` composable and delay the initial queries until the value is found.

### Steps to test

_In order to test this PR, you will currently need to have the Cypress Cloud running locally or point to Cloud staging_

* Open Cypress and log in to your local Cloud instance
* Open a project that has recorded runs
* Verify that the Debug page and the sidebar badge represent the appropriate "relevant" run based on the algorithm defined above and documented in the code in `RelevantRunsDataSource`
* Start recording a new run to this project
* Verify that the new RUNNING run is shown in the banner on the Debug page as running (banner implemented here https://github.com/cypress-io/cypress/issues/24853)
* When the run completes, verify that the banner updates to show the completed status and a "View run" link (banner implemented here https://github.com/cypress-io/cypress/issues/24853)
* Click the "View run" and verify that the latest run is now shown as the current run on the Debug page

### How has the user experience changed?

N/A Backend changes only

### PR Tasks

- [ ] Have tests been added/updated?
- [ ] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

**Comment by ZachJW34** [comment]
Noticed a lot of network requests and thrashing of the UI which I believe is related to how the queries are hooked up and their reactivity. I'm on the latest cloud feature branch with the latest changes.


https://user-images.githubusercontent.com/25158820/210858197-f1e15268-cc25-4b21-a80b-66a2853f83a2.mov

File: packages/app/src/pages/Debug.vue
Comment (marktnoonan) [inline packages/app/src/pages/Debug.vue]: It might be a bit more intuitive to have `variables` be a computed value, and watch only `variables.runNumber` to trigger `query.executeQuery()` when needed?

File: packages/data-context/src/sources/RelevantRunsDataSource.ts
Comment (marktnoonan) [inline packages/data-context/src/sources/RelevantRunsDataSource.ts]:
```suggestion
        // use it as the current run
```

File: packages/data-context/test/unit/sources/RelevantRunsDataSource.spec.ts
Comment (marktnoonan) [inline packages/data-context/test/unit/sources/RelevantRunsDataSource.spec.ts]:
```suggestion
```

File: packages/data-context/test/unit/sources/RelevantRunsDataSource.spec.ts
Comment (marktnoonan) [inline packages/data-context/test/unit/sources/RelevantRunsDataSource.spec.ts]: Could we track this with an issue number, or is this still a TODO before this PR is merged?

**Comment by marktnoonan** [comment]
Is this state expected? No badge in left nav, project is connected but hadn't recorded a run yet.

Screen Shot 2023-01-05 at 3 41 49 PM

First run experience looks good after refresh. "New" badge reappeared after the same refresh.

Screen Shot 2023-01-05 at 3 46 46 PM

File: packages/graphql/src/schemaTypes/objectTypes/gql-CurrentProject.ts
Comment (ZachJW34) [inline packages/graphql/src/schemaTypes/objectTypes/gql-CurrentProject.ts]: Does the frontend ever have to know about the relevant run numbers? Seems like we could handle more on the backend. Right now we:
* Frontend query for relevant runs
* Backend request to cloud to get relevant runs
* Frontend query for run information based on relevant runs
* Backend request to cloud to get run information.

Can the frontend just ask for relevant runs and be given the full run info for next/current? We could skip a cycle

File: packages/graphql/src/schemaTypes/objectTypes/gql-CurrentProject.ts
Comment (ZachJW34) [inline packages/graphql/src/schemaTypes/objectTypes/gql-CurrentProject.ts:258]: Should this follow the pattern for `cloudSpec` and utilize the `remoteField`, allowing this resolver to be deferred? I don't think we want the frontend query to stall based on a cloud query.

File: packages/app/src/composables/useTestsForDebug.ts
Comment (ZachJW34) [inline packages/app/src/composables/useTestsForDebug.ts:16]: This is very clunky compared to a mutation, but I don't see any alternative other than using a mutation.

File: packages/app/src/composables/useTestsForDebug.ts
Comment (ZachJW34) [inline packages/app/src/composables/useTestsForDebug.ts]: Does this need to be computed? Something feels strange about a composable returning a function that returns reactive values.

File: packages/app/src/navigation/SidebarNavigation.vue
Comment (ZachJW34) [inline packages/app/src/navigation/SidebarNavigation.vue:190]: This was to cover any error status like 'OVERLIMIT' and others. Are we not supposed to show the badge for these cases?

File: packages/app/src/navigation/SidebarNavigationContainer.vue
Comment (ZachJW34) [inline packages/app/src/navigation/SidebarNavigationContainer.vue:42]: We shouldn't use `watchEffect` here, as `query` is reactive and you don't want this re-executing when query updates. This should probably just be `watch(relevantRuns, () => {...})`

File: packages/data-context/src/sources/RelevantRunsDataSource.ts
Comment (ZachJW34) [inline packages/data-context/src/sources/RelevantRunsDataSource.ts:122]: Nit: I'd rather see a `result.data.cloudProjectBySlug.runsByCommitShas.reduce` or `map + filter` combo than utilize lodash methods. Seeing `.map().filter().first()` below threw me for a _loop_, and I think writing this with plain ol js would be much more readable.

Comment (warrensplayer) [reply packages/data-context/src/sources/RelevantRunsDataSource.ts:122]: Using lodash `chain` to leverage the lazy evaluation and shortcut abilities: https://lodash.com/docs/4.17.15#lodash

File: packages/app/src/navigation/SidebarNavigationContainer.vue
Comment (warrensplayer) [reply packages/app/src/navigation/SidebarNavigationContainer.vue:42]: I believe this is okay as written. I am pretty sure the `query` itself is not reactive and does not change here.  It is set just the one time as the result of `useQuery`.  The underlying fields of query, like `data`, are reactive.  In this case, I want the `query.executeQuery()` to run if the `variables` change.

File: packages/graphql/src/schemaTypes/objectTypes/gql-CurrentProject.ts
Comment (warrensplayer) [reply packages/graphql/src/schemaTypes/objectTypes/gql-CurrentProject.ts]: Due to limitations on how we can delegate queries to the cloud, this is the best way I could find to do this.  Since the `runByNumber` field required a variable and it is nested under the `cloudRunBySlug`, the current code cannot easily create an additional nested field that delegates to it. I played around with several methods.  It should be possible, but I had to move on.

File: packages/graphql/src/schemaTypes/objectTypes/gql-CurrentProject.ts
Comment (warrensplayer) [reply packages/graphql/src/schemaTypes/objectTypes/gql-CurrentProject.ts:258]: I do not need this anymore.  I am using a subscription to get this data since it needs to refresh with a poller

File: packages/app/src/composables/useTestsForDebug.ts
Comment (warrensplayer) [reply packages/app/src/composables/useTestsForDebug.ts]: No, it does not need to be a computed

File: packages/graphql/src/schemaTypes/objectTypes/gql-Mutation.ts
Comment (mike-plummer) [inline packages/graphql/src/schemaTypes/objectTypes/gql-Mutation.ts]: I feel like this event should be emitted from within `ctx.relevantRuns.moveToNext()`

File: packages/graphql/src/schemaTypes/objectTypes/gql-Subscription.ts
Comment (mike-plummer) [inline packages/graphql/src/schemaTypes/objectTypes/gql-Subscription.ts]: Is this description accurate?

File: packages/graphql/src/schemaTypes/objectTypes/gql-CurrentProject.ts
Comment (mike-plummer) [inline packages/graphql/src/schemaTypes/objectTypes/gql-CurrentProject.ts]: Is this a follow-on task? M2?

File: packages/data-context/src/sources/RelevantRunSpecsDataSource.ts
Comment (mike-plummer) [inline packages/data-context/src/sources/RelevantRunSpecsDataSource.ts]: Address this?

**Review by mike-plummer — APPROVED** [review]
Is working great locally minus the issues I pointed out on Slack. Just a couple comments and it looks like there's a couple from Zach still pending as well, will approve assuming they are answered

File: packages/app/src/navigation/SidebarNavigation.vue
Comment (warrensplayer) [reply packages/app/src/navigation/SidebarNavigation.vue:190]: I rearranged to make sure those cases were covered.

File: packages/app/src/composables/useTestsForDebug.ts
Comment (warrensplayer) [reply packages/app/src/composables/useTestsForDebug.ts:16]: This is just the way you have to do it.

File: packages/data-context/test/unit/sources/RelevantRunsDataSource.spec.ts
Comment (warrensplayer) [reply packages/data-context/test/unit/sources/RelevantRunsDataSource.spec.ts]: Tests added

File: packages/data-context/test/unit/sources/RelevantRunsDataSource.spec.ts
Comment (warrensplayer) [reply packages/data-context/test/unit/sources/RelevantRunsDataSource.spec.ts]: comments removed

File: packages/graphql/src/schemaTypes/objectTypes/gql-Mutation.ts
Comment (warrensplayer) [reply packages/graphql/src/schemaTypes/objectTypes/gql-Mutation.ts]: agreed!

File: packages/data-context/src/sources/RelevantRunSpecsDataSource.ts
Comment (warrensplayer) [reply packages/data-context/src/sources/RelevantRunSpecsDataSource.ts]: Handled

File: packages/graphql/src/schemaTypes/objectTypes/gql-Subscription.ts
Comment (warrensplayer) [reply packages/graphql/src/schemaTypes/objectTypes/gql-Subscription.ts]: Updated

**Comment by marktnoonan** [comment]
I caught a brief flash of "new" here before this resolved to 0. Not blocking but possibly we could have a little delay to make sure nothing is coming from the cloud before we show the "new" status.

Screen Shot 2023-01-17 at 11 23 24 AM

**Comment by warrensplayer** [comment]
> I caught a brief flash of "new" here before this resolved to 0. Not blocking but possibly we could have a little delay to make sure nothing is coming from the cloud before we show the "new" status.

@marktnoonan Yes, there are some scenarios causing "flashes" of content.  I am going to file a follow up issue for that.

**Comment by marktnoonan** [comment]
What is the expectation when more than one run is pending? We just see the newest in this page?

Screen Shot 2023-01-17 at 11 27 18 AM

File: packages/app/src/debug/DebugFailedTest.vue
Comment (marktnoonan) [inline packages/app/src/debug/DebugFailedTest.vue:14]: good catch

**Review by marktnoonan — APPROVED** [review]
Tested as described in the PR and everything works. I have a few questions and captured them on the fly here: https://www.loom.com/share/bd58758e23d8469d9300dbe5fb37288d - it's a long one but might be good for reference.

~There are a few instances where state flickers and I think some strategic debounces or waits could help us out as we polish.~ Nvm, I saw our earlier convo that we'll make a new issue.

The main usability issue I see at the the moment is the need to manually click a button called "View Run" (which is different to the other button "View in Cypress Cloud") on the just-finished run in order to "promote it" to be the run that you are actively debugging.

It feels weird that when Cypress "knows about a run" in the Debug page, it's possible for the run's page not to know about that run yet, and that there's no way to click through to the Cloud from the pending run. I think this has to do with the fact that we don't poll in the background when the user is not on the runs page, and we don't immediately poll again when they come back to it.

None of these are blocking issues for this PR, just a head start on feedback on the feature overall I guess.

**Comment by warrensplayer** [comment]
Merging this to the feature branch.  Items left to complete are filed here:

https://github.com/cypress-io/cypress/issues/25486
https://github.com/cypress-io/cypress/issues/25487

File: packages/graphql/src/schemaTypes/objectTypes/gql-CurrentProject.ts
Comment (warrensplayer) [reply packages/graphql/src/schemaTypes/objectTypes/gql-CurrentProject.ts]: Will be addressed here: https://github.com/cypress-io/cypress/issues/25487

**Pull request merged**
